mlir/lib/Catalyst/Transforms/SplitMultipleTapes.cpp

```diff
@@ -0,0 +1,307 @@
+        for (Value v : RetValues) {
+            RetTypes.push_back(v.getType());
+        }
+
+        // 2. Create a scf.execute_region and wrap it around the tape ops
+        // The scf.execute_region needs to yield (aka return) the values identified
+        std::pair<scf::ExecuteRegionOp, scf::YieldOp> WrappingRegionAndYield =
+            wrapTapeOpsInSCFRegion(TapeOps, RetValues, RetTypes, builder,
+                                   OriginalMultitapeFunc->getLoc());
+        scf::ExecuteRegionOp executeRegionOp = WrappingRegionAndYield.first;
+        scf::YieldOp SCFRegionYieldOp = WrappingRegionAndYield.second;
+
+        // 3. The scf outliner automatically captures values not defined in the
+        // to-be-outlined function body and transforms them to the outlined function's
+        // arguments, but it does not automatically propagate the returned value
+        // downstream to replace the previously used values. This we need to do manually.
+
+        // At this stage, the scf.yield values ARE the ones to be replaced by
+        // the scf region's result values, since we never replaced anything yet
+        // Of course, don't replace the use in the region itself!
+        propagateSCFRetValsDownstream(executeRegionOp, SCFRegionYieldOp, TapeOps, RetValues);
+
+        // 4. Outline the region
+        func::CallOp call;
+        FailureOr<func::FuncOp> outlined = outlineSingleBlockRegion(
+            builder, OriginalMultitapeFunc->getLoc(), executeRegionOp.getRegion(),
+            OriginalMultitapeFunc.getSymName().str() + "_tape_" + std::to_string(tape_number),
+            &call);
+        OutlinedFuncs.push_back(outlined);
+
+        if (failed(outlined)) {
+            return failure();
+        }
+        return success();
+    } // createTapeFunction()
+
+    void runOnOperation() override
+    {
+        Operation *module = getOperation();
+        mlir::IRRewriter builder(module->getContext());
+
+        // 1. Identify the functions with multiple tapes
+        // Walk through each function and count the number of devices.
+        // In frontend when tracing (jax_tracers.py/trace_quantum_function),
+        // each tape has its own quantum.device operation attached to it
+        SmallVector<func::FuncOp> MultitapePrograms;
+        module->walk([&](func::FuncOp func) {
+            if (countTapes(func) >= 2) {
+                MultitapePrograms.push_back(func);
+            }
+        });
+
+        // Do nothing and exit for classical and single-tape programs
+        if (MultitapePrograms.empty()) {
+            return;
+        }
+
+        // 2. Parse the function into tapes
+        // total number of operation lists is number of tapes +2
+        // for classical pre and post processing
+        for (func::FuncOp func : MultitapePrograms) {
+            SmallVector<std::shared_ptr<SmallVector<Operation *>>> OpsEachTape(countTapes(func) + 2,
```

Comment:
I don't know if there is much of a difference between a SmallVector and a Vector in cases that exceed the static storage, but I don't think I would expect the number of instructions in a tape to be "small" 😅 

But maybe it's safe to say "always use a small vector"?

Reply:
I think Smallvector just becomes std::vector when it's not actually small. 

In any case, LLVM does prefer smallvector over std vector: https://llvm.org/docs/ProgrammersManual.html#llvm-adt-smallvector-h. However for cases where we know the vector won't be small I agree it does make more sense to just use std::vector.

Reply:
Interesting, it does pay to read documentation 😅 

> On the other hand, this makes the size of the SmallVector itself large, so you don’t want to allocate lots of them (doing so will waste a lot of space). As such, SmallVectors are most useful when on the stack.

It sounds like since we are nesting the vectors that the "allocating lots of them" regime would apply, and they are not on the stack either.
The nested small vector structure also means the outer vector would be super large, since it holds some amount of inner vectors in its static storage 😅 
`SmallVector<Vector>` might be the way to go here (it's reasonable that the number of tapes could be small).

Reply:
The "outer vector itself would be very large" problem is why I originally opted for `Smallvector<SmallVector*>`, i.e. the outer vector would only store tiny handles to the inner vectors (which themselves live somewhere else). 

Agreed that the inner vector shouldn't be a smallvector since number of instructions in an outlined function usually will be big.